Simplify str_to_display() and its uses

jdufresne opened the pull request:
With Python 2 dropped, only bytes need to be handled. Passing str was a
nop. These bytes/str boundaries are much clearer now.

There remains one use of str_to_display(): constole_to_str(). It
converts subprocess bytes output to str while handling decoding errors.

jdufresne commented:
This could be simplifies further by using the subprocess [`encoding` and `errors` arguments](https://docs.python.org/3/library/subprocess.html#subprocess.run). We could then remove `console_to_str` and `str_to_display`. The difference would be that, in the event of a decoding error, no warning would be logged. Otherwise, the user facing behavior would be the same. Thoughts?

pradyunsg commented:
The only place this is used is Python 3, correct? I'd check the git blame to cross-reference what the history is on this function, and if it is indeed possible to (1) simplify code and (2) have tests to trust that we don't need to look back at this later, that'd be great!

jdufresne commented:
Thanks for the feedback.

> The only place this is used is Python 3, correct?

Sorry, I'm not quite clear on what you're asking, but I'm happy to research what you're looking for. All code on `master` is Python 3 only, right? So, I presume, "yes"...?

This code is used to format console bytes output for logging. It handles the bytes-to-str decoding errors by using "backslashreplace". Starting with Python 3.6, the `subprocess` module supports decoding and decoding error handling, so "backslashreplace" could be passed directly to the module functions. The only difference is, in the event of a decoding error, the "backslashreplace" would automatically kick in without the chance for pip to log a warning. Is this lack of warning acceptable?

> I'd check the git blame to cross-reference what the history is on this function

The `str_to_display()` function was introduced in PR #6671 (2019-07-02).

The `console_to_str()` function was introduced in PR #4486 (2017-05-27).

> have tests to trust that we don't need to look back at this later, that'd be great!

Some tests already exist, but I added a new explicit one:

`test_unicode_decode_error` in `tests/unit/test_utils_subprocess.py`.

This test shows that in the event of a `UnicodeDecode` error two things happen:

1. A warning is logged "Subprocess output does not appear to be encoded as UTF-8"
2. The stdout is logged with errors="backslashreplace"

Perhaps it would be easier to show my suggestion with code. I have split this PR out into 3 commits:

1. Add test, this shows the "before" behavior described
2. The simplified version (original PR), shows the same "before" behavior
3. The suggestion to further simplify by removing these compat functions

Commit 3 demonstrates the change in behavior I'm describing. The "Subprocess output does not appear to be encoded as UTF-8" warning is no longer shown, but the stdout is still logged using "backslashescape".

I can easily dropped the third commit if it is not desirable, but I think the reduction in compat code is nice and the trade off worth it for nearly the same result. In normal usage, I expect a subprocess decoding error to be rare.

pradyunsg commented:
Ah! Right. Well, I genuinely appreciate you breaking up the commits! It certainly helped me understand what you said. 

I agree that this is a really nice simplification, and IIUC, the behaviour change here should be fine. I'll let @pfmoore or @uranusjr confirm if this change is indeed good enough though, since my understanding is that Windows was the platform where this encoding management mattered the most during PEP 517's initial rollout.

uranusjr commented on src/pip/_internal/utils/subprocess.py:
```diff
@@ -191,8 +186,7 @@ def call_subprocess(
     while True:
-        # The "line" value is a unicode string in Python 2.
-        line = console_to_str(proc.stdout.readline())
+        line = proc.stdout.readline()
```
Probably worth a `# type: str` and maybe a comment so future readers know the type of `line` is `str` and why (because we specified `errors`).

jdufresne replied:
Makes sense to me. Added in the latest revision.

uranusjr commented on src/pip/_internal/vcs/versioncontrol.py:
```diff
@@ -130,10 +130,9 @@ def call_subprocess(
         line = None
         if proc.stdout:
-            line = console_to_str(proc.stdout.readline())
+            line = proc.stdout.readline()
```
Same here.
